packages/block-library/src/comments/edit/comments-legacy.js

```diff
@@ -0,0 +1,128 @@
+/**
+ * External dependencies
+ */
+import classnames from 'classnames';
+
+/**
+ * WordPress dependencies
+ */
+import {
+	AlignmentControl,
+	BlockControls,
+	Warning,
+	useBlockProps,
+	store as blockEditorStore,
+} from '@wordpress/block-editor';
+import { __, sprintf } from '@wordpress/i18n';
+import { useSelect, useDispatch } from '@wordpress/data';
+import { useEntityProp, store as coreStore } from '@wordpress/core-data';
+import { createBlock } from '@wordpress/blocks';
+import { Button } from '@wordpress/components';
+
+/**
+ * Internal dependencies
+ */
+import Placeholder from './placeholder';
+
+export default function CommentsLegacy( {
+	attributes,
+	setAttributes,
+	context: { postType, postId },
+	clientId,
+} ) {
+	const { textAlign } = attributes;
+
+	const [ commentStatus ] = useEntityProp(
+		'postType',
+		postType,
+		'comment_status',
+		postId
+	);
+
+	const { defaultCommentStatus } = useSelect(
+		( select ) =>
+			select( blockEditorStore ).getSettings()
+				.__experimentalDiscussionSettings
+	);
+
+	const isSiteEditor = postType === undefined || postId === undefined;
+
+	const postTypeSupportsComments = useSelect( ( select ) =>
+		postType
+			? !! select( coreStore ).getPostType( postType )?.supports.comments
+			: false
+	);
+
+	let warning = __(
+		"Comments block: You're currently using this block in legacy mode. The following is just a placeholder, not a real comment. The final styling may differ because it also depends on the current theme. For better compatibility with the Block Editor, please consider switching the block to its editable mode."
+	);
+	let showPlacholder = true;
+
+	if ( ! isSiteEditor && 'open' !== commentStatus ) {
+		if ( 'closed' === commentStatus ) {
+			warning = sprintf(
+				/* translators: 1: Post type (i.e. "post", "page") */
+				__(
+					'Post Comments block: Comments to this %s are not allowed.'
+				),
+				postType
+			);
+			showPlacholder = false;
+		} else if ( ! postTypeSupportsComments ) {
+			warning = sprintf(
+				/* translators: 1: Post type (i.e. "post", "page") */
+				__(
+					'Post Comments block: Comments for this post type (%s) are not enabled.'
+				),
+				postType
+			);
+			showPlacholder = false;
+		} else if ( 'open' !== defaultCommentStatus ) {
+			warning = __( 'Post Comments block: Comments are not enabled.' );
+			showPlacholder = false;
+		}
+	}
+
+	const { replaceBlock } = useDispatch( blockEditorStore );
+
+	const removeLegacy = () => {
+		const { legacy, ...attributesWithoutLegacy } = attributes;
+		replaceBlock(
+			clientId,
+			createBlock( 'core/comments', attributesWithoutLegacy )
+		);
```

Comment:
I agree that the [3 undo levels](https://github.com/WordPress/gutenberg/pull/41807#issuecomment-1180747046) aren't super-intuitive 😕 I _think_ that at least two of those result from `createBlock` and `replaceBlock`? That's what going back with Undo seems to indicate, anyway.

Can we maybe avoid that by removing the `legacy` attribute via `setAttribute` here (and, if needed, inserting the child blocks from `TEMPLATE`)?

Reply:
The good news: Just setting the attribute to `false` seems to be enough to convert the block (the child blocks are inserted automatically).

```diff
diff --git a/packages/block-library/src/comments/block.json b/packages/block-library/src/comments/block.json
index 151664b290..042184b9b8 100644
--- a/packages/block-library/src/comments/block.json
+++ b/packages/block-library/src/comments/block.json
@@ -12,7 +12,8 @@
 			"default": "div"
 		},
 		"legacy": {
-			"type": "boolean"
+			"type": "boolean",
+			"default": false
 		}
 	},
 	"supports": {
diff --git a/packages/block-library/src/comments/edit/comments-legacy.js b/packages/block-library/src/comments/edit/comments-legacy.js
index cf83d6c033..ee51d3e4b4 100644
--- a/packages/block-library/src/comments/edit/comments-legacy.js
+++ b/packages/block-library/src/comments/edit/comments-legacy.js
@@ -14,9 +14,8 @@ import {
 	store as blockEditorStore,
 } from '@wordpress/block-editor';
 import { __, sprintf } from '@wordpress/i18n';
-import { useSelect, useDispatch } from '@wordpress/data';
+import { useSelect } from '@wordpress/data';
 import { useEntityProp, store as coreStore } from '@wordpress/core-data';
-import { createBlock } from '@wordpress/blocks';
 import { Button } from '@wordpress/components';
 
 /**
@@ -28,7 +27,6 @@ export default function CommentsLegacy( {
 	attributes,
 	setAttributes,
 	context: { postType, postId },
-	clientId,
 } ) {
 	const { textAlign } = attributes;
 
@@ -83,14 +81,8 @@ export default function CommentsLegacy( {
 		}
 	}
 
-	const { replaceBlock } = useDispatch( blockEditorStore );
-
 	const removeLegacy = () => {
-		const { legacy, ...attributesWithoutLegacy } = attributes;
-		replaceBlock(
-			clientId,
-			createBlock( 'core/comments', attributesWithoutLegacy )
-		);
+		setAttributes( { legacy: false } );
 	};
 
 	const actions = [

```

The bad news: Still 3 undo levels 😅

Reply:
Sure, I'll try that. 🙂